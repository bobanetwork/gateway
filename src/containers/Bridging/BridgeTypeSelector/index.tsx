import { setBridgeType } from 'actions/bridgeAction'
import { setNetwork } from 'actions/networkAction'
import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import {
  selectActiveNetworkType,
  selectBridgeType,
  selectNetwork,
  selectNetworkType,
} from 'selectors'
import {
  DEFAULT_NETWORK,
  NetworkType,
  networkLimitedAvailability,
} from 'util/network/network.util'
import { BridgeTabItem, BridgeTabs } from './style'

export enum BRIDGE_TYPE {
  CLASSIC = 'CLASSIC',
  FAST = 'FAST',
  LIGHT = 'LIGHT',
  THIRD_PARTY = 'THIRD_PARTY',
}

const BridgeTypeSelector = () => {
  const dispatch = useDispatch<any>()
  const bridgeType = useSelector(selectBridgeType())
  const networkType = useSelector(selectNetworkType())
  const network = useSelector(selectNetwork())
  const isOnLimitedNetwork = networkLimitedAvailability(networkType, network)
  const activeNetworkType = useSelector(selectActiveNetworkType())

  const isMainnet =
    useSelector(selectActiveNetworkType()) === NetworkType.MAINNET

  const switchToFullySupportedNetwork = () => {
    // change network back to fully supported network when leaving light bridge
    dispatch(
      setNetwork({
        network: DEFAULT_NETWORK.chain,
        name: DEFAULT_NETWORK.name,
        networkIcon: DEFAULT_NETWORK.icon,
        chainIds: DEFAULT_NETWORK.chainId,
        networkType,
      })
    )
  }

  useEffect(() => {
    if (bridgeType !== BRIDGE_TYPE.LIGHT && isOnLimitedNetwork) {
      switchToFullySupportedNetwork()
    }
  }, [bridgeType, isOnLimitedNetwork])

  const onTabClick = (payload: any) => {
    dispatch(setBridgeType(payload))
  }

  useEffect(() => {
    dispatch(setBridgeType(BRIDGE_TYPE.CLASSIC))
  }, [activeNetworkType])

  return (
    <BridgeTabs>
      <BridgeTabItem
        data-testid="classic-btn"
        active={bridgeType === BRIDGE_TYPE.CLASSIC}
        onClick={() => onTabClick(BRIDGE_TYPE.CLASSIC)}
      >
        Classic
      </BridgeTabItem>

      <BridgeTabItem
        data-testid="light-btn"
        active={bridgeType === BRIDGE_TYPE.LIGHT}
        onClick={() => onTabClick(BRIDGE_TYPE.LIGHT)}
      >
        Light
      </BridgeTabItem>

      {isMainnet ? (
        <BridgeTabItem
          data-testid="third-party-btn"
          active={bridgeType === BRIDGE_TYPE.THIRD_PARTY}
          onClick={() => onTabClick(BRIDGE_TYPE.THIRD_PARTY)}
        >
          Third Party
        </BridgeTabItem>
      ) : null}
    </BridgeTabs>
  )
}

export default BridgeTypeSelector
