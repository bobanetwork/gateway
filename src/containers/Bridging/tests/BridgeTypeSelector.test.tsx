import { render } from '@testing-library/react'
import React from 'react'
import CustomThemeProvider from 'themes'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import BridgeTypeSelector from '../BridgeTypeSelector'

const mockStore = configureStore()

const renderBrdigeTypeSelector = ({ options = null }: any) => {
  return render(
    <Provider
      store={mockStore({
        ui: {
          theme: 'dark',
        },
        network: {
          networkType: 'Mainnet',
        },
        bridge: {
          bridgeType: 'LIGHT',
        },
        setup: {
          accountEnabled: false,
          netLayer: true,
          baseEnabled: false,
          walletAddress: '0x1e2855A0EA33d5f293E5Ba2018874FAB9a7F05B3',
        },
        ...options,
      })}
    >
      <CustomThemeProvider>
        <BridgeTypeSelector></BridgeTypeSelector>
      </CustomThemeProvider>
    </Provider>
  )
}

describe('Testing BridgeTypeSelector', () => {
  beforeAll(() => {
    jest.useFakeTimers().setSystemTime(new Date(2023, 7, 24))
  })
  afterAll(() => {
    jest.useRealTimers()
  })

  test('Test BridgeTypeSelector renders', () => {
    const { asFragment } = renderBrdigeTypeSelector({})
    expect(asFragment()).toMatchSnapshot()
  })
})
