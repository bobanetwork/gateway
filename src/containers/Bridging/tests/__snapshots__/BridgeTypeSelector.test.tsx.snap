// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing BridgeTypeSelector Test BridgeTypeSelector renders 1`] = `
<DocumentFragment>
  <div
    class="sc-aXZVg fRVoBl"
  >
    <div
      class="sc-gEvEer bGdtCD"
    >
      Classic
    </div>
    <div
      class="sc-gEvEer bGdtCD"
    >
      Fast
    </div>
  </div>
</DocumentFragment>
`;
